components/bpmn/org.wso2.carbon.bpmn.analytics.publisher/src/main/java/org/wso2/carbon/bpmn/analytics/publisher/utils/AnalyticsPublishServiceUtils.java

```diff
@@ -0,0 +1,301 @@
+				resource.setProperty(AnalyticsPublisherConstants.LAST_PROCESS_INSTANCE_PUBLISH_TIME,
+				                     String.valueOf(lastProcessInstancePublishTime));
+				registry.put(AnalyticsPublisherConstants.PROCESS_RESOURCE_PATH, resource);
+			}
+			log.debug("End of writing last completed process instance publish time...");
+		} catch (RegistryException e) {
+			String errMsg = "Registry error while writing the process instance publish time.";
+			log.error(errMsg, e);
+		}
+	}
+
+	/**
+	 * Write last completed task instance end time to carbon registry
+	 *
+	 * @param historicTaskInstanceList List of historic task instances
+	 */
+	private void writeTaskEndTimeToRegistry(List<HistoricTaskInstance> historicTaskInstanceList) {
+		log.debug("Start writing last completed task instance end time...");
+		Date lastTaskInstanceDate =
+				historicTaskInstanceList.get(historicTaskInstanceList.size() - 1).getEndTime();
+		try {
+			PrivilegedCarbonContext privilegedContext =
+					PrivilegedCarbonContext.getThreadLocalCarbonContext();
+			Registry registry = privilegedContext.getRegistry(RegistryType.SYSTEM_GOVERNANCE);
+			Resource resource;
+			if (!registry.resourceExists(AnalyticsPublisherConstants.TASK_RESOURCE_PATH)) {
+				resource = registry.newResource();
+				resource.addProperty(AnalyticsPublisherConstants.LAST_TASK_INSTANCE_END_TIME,
+				                     String.valueOf(lastTaskInstanceDate));
+				registry.put(AnalyticsPublisherConstants.TASK_RESOURCE_PATH, resource);
+			} else {
+				resource = registry.get(AnalyticsPublisherConstants.TASK_RESOURCE_PATH);
+				resource.setProperty(AnalyticsPublisherConstants.LAST_TASK_INSTANCE_END_TIME,
+				                     String.valueOf(lastTaskInstanceDate));
+				registry.put(AnalyticsPublisherConstants.TASK_RESOURCE_PATH, resource);
+			}
+			log.debug("End of writing last completed task instance end time...");
+		} catch (RegistryException e) {
+			String errMsg = "Registry error while writing the task instance end time.";
+			log.error(errMsg, e);
+		}
+	}
+
+	/**
+	 * Read last completed process/task instance end time from carbon registry
+	 *
+	 * @param resourcePath resource path in the carbon registry
+	 * @param propertyPath property path in the resource
+	 * @return the end time of last completed process/task instance
+	 */
+
+	private String readPublishTimeFromRegistry(String resourcePath, String propertyPath) {
+		String time = null;
+		try {
+			PrivilegedCarbonContext carbonContext =
+					PrivilegedCarbonContext.getThreadLocalCarbonContext();
+			Registry registry = carbonContext.getRegistry(RegistryType.SYSTEM_GOVERNANCE);
+			if (registry != null) {
+				if (registry.resourceExists(resourcePath)) {
+					log.debug("Process instance resource path exists...");
+					Resource resource = registry.get(resourcePath);
+					time = resource.getProperty(propertyPath);
+				}
+			} else {
+				log.debug("Registry is null...");
```

Comment:
If the registry is null, can we continue. If not, we have to throw an exception.